module/VuFind/src/VuFind/Controller/SearchApiController.php

```diff
@@ -0,0 +1,626 @@
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA    02111-1307    USA
+ *
+ * @category VuFind
+ * @package  Controller
+ * @author   Ere Maijala <ere.maijala@helsinki.fi>
+ * @license  http://opensource.org/licenses/gpl-2.0.php GNU General Public License
+ * @link     https://vufind.org/wiki/development:plugins:controllers Wiki
+ */
+namespace VuFind\Controller;
+
+use VuFind\I18n\TranslatableString;
+
+/**
+ * SearchApiController Class
+ *
+ * Controls the Search API functionality
+ *
+ * @category VuFind
+ * @package  Controller
+ * @author   Ere Maijala <ere.maijala@helsinki.fi>
+ * @license  http://opensource.org/licenses/gpl-2.0.php GNU General Public License
+ * @link     https://vufind.org/wiki/development:plugins:controllers Wiki
+ */
+class SearchApiController extends \VuFind\Controller\AbstractSearch
+    implements ApiInterface
+{
+    use ApiTrait;
+
+    /**
+     * Permission required for the record endpoint
+     *
+     * @var string
+     */
+    protected $recordAccessPermission = 'access.api.Record';
+
+    /**
+     * Permission required for the search endpoint
+     *
+     * @var string
+     */
+    protected $searchAccessPermission = 'access.api.Search';
+
+    /**
+     * Record action
+     *
+     * @return \Zend\Http\Response
+     */
+    public function recordAction()
+    {
+        // Disable session writes
+        $this->disableSessionWrites();
+
+        $this->determineOutputMode();
+
+        if ($result = $this->isAccessDenied($this->recordAccessPermission)) {
+            return $result;
+        }
+
+        $request = $this->getRequest()->getQuery()->toArray()
+            + $this->getRequest()->getPost()->toArray();
+
+        if (!isset($request['id'])) {
+            return $this->output([], self::STATUS_ERROR, 400, 'Missing id');
+        }
+
+        $requestedFields = $this->getFieldList($request);
+
+        $loader = $this->getServiceLocator()->get('VuFind\RecordLoader');
+        try {
+            if (is_array($request['id'])) {
+                $results = $loader->loadBatchForSource($request['id']);
+            } else {
+                $results[] = $loader->load($request['id']);
+            }
+        } catch (\Exception $e) {
+            return $this->output(
+                [], self::STATUS_ERROR, 400,
+                'Error loading record'
+            );
+        }
+
+        $records = [];
+        foreach ($results as $result) {
+            $records[] = $this->getFields($result, $requestedFields);
+        }
+
+        $this->filterArrayValues($records);
+
+        $response = [
+            'resultCount' => count($results)
+        ];
+        if ($records) {
+            $response['records'] = $records;
+        }
+
+        return $this->output($response, self::STATUS_OK);
+    }
+
+    /**
+     * Search action
+     *
+     * @return \Zend\Http\Response
+     */
+    public function searchAction()
```

Comment:
...in general, I wonder if we should introduce a \VuFind\API namespace and put a bunch of helper classes there to make the controllers thinner. Obviously it makes sense for the controllers to handle the high-level framework stuff of constructing responses and parsing input, but anything related to formatting API responses, etc., might be better positioned in its own classes (which would enable better testing, greater separation from framework-specific details, etc., etc.)